rust/kernel/kasync/executor.rs

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Kernel support for executing futures.
+
+use crate::{
+    linked_list::{GetLinks, GetLinksWrapped, Links},
+    sync::{Ref, RefBorrow},
+    types::PointerWrapper,
+    Result,
+};
+use core::future::Future;
+use core::task::{Context, Poll, RawWaker, RawWakerVTable, Waker};
+
+pub mod local;
+
+/// An environment for executing tasks.
+pub trait Executor {
+    /// Starts executing a task defined by the given future.
+    fn spawn(self: RefBorrow<'_, Self>, future: impl Future + 'static + Send) -> Result;
+}
+
+struct TaskListAdapter;
+impl GetLinks for TaskListAdapter {
+    type EntryType = dyn Task;
+    fn get_links(data: &dyn Task) -> &Links<dyn Task> {
+        data.get_links()
+    }
+}
+
+impl GetLinksWrapped for TaskListAdapter {
+    type Wrapped = Ref<dyn Task>;
+}
+
+trait Task: Send {
+    /// Returns a waker for this task.
+    fn waker(self: Ref<Self>) -> Waker;
+
+    /// Polls the task.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that only one thread at a time calls this function.
```

Comment:
Is it safe to call reentrantly?

Reply:
No. What interpretation of "only one thread at a time calls this function" would allow for a reentrant call? (I really mean this question, and will adapt the wording to make it clear.)

Reply:
Some functions would cause a data race when called from multiple threads due to using plain writes with UnsafeCell, but are completely fine to reentrantly call from within the same thread as there would be no data race in that case.

https://users.rust-lang.org/t/a-better-term-than-thread-safe/73199/4 is somewhat relevant I think.